core/common/src/main/java/com/google/jetpackcamera/core/common/RefCounted.kt

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (C) 2024 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.jetpackcamera.core.common
+
+import android.util.Log
+import kotlinx.atomicfu.atomic
+import kotlinx.atomicfu.loop
+
+/**
+ * A thread-safe, lock-free class for managing the lifecycle of an object using ref-counting.
+ *
+ * The object that is ref-counted can be initialized late using [initialize].
+ *
+ * @param[debugRefCounts] whether to print debug log statements.
+ * @param[onRelease] a block that will be invoked once when the ref-count reaches 0.
+ */
+class RefCounted<T : Any>(
+    private val debugRefCounts: Boolean = false,
+    private val onRelease: (T) -> Unit = {}
+) {
+    private val refCounted = atomic(uninitialized<T>())
+
+    /**
+     * Initializes the ref-count managed object with the object being managed.
+     *
+     * This also initializes the implicit ref-count to 1.
+     *
+     */
+    fun initialize(newValue: T) {
+        val initialVal = Pair(newValue, 1)
+        check(refCounted.compareAndSet(uninitialized(), initialVal)) {
+            "Ref-count managed object has already been initialized."
+        }
+
+        if (debugRefCounts) {
+            Log.d(
+                TAG,
+                "RefCounted@${"%x".format(hashCode())}<${newValue::class.simpleName}> " +
+                    "initialized: [refCount: 1, value: $newValue]",
+                Throwable()
+            )
+        }
+    }
+
+    /**
+     * Retrieves the underlying managed object, increasing the ref-count by 1.
+     *
+     * This increases the ref-count if the object has not already been released. If the object has
+     * been released, `null` is returned.
+     *
+     * All calls to this function must be paired with [release], unless `null` is returned.
+     */
+    fun acquire(): T? {
+        check(refCounted.value != uninitialized<T>()) {
+            "Ref-count managed object has not yet been initialized. Unable to acquire."
```

Comment:
nit: "Unable to release."

Reply:
Good catch. Made this change in the `release()` function.